client/client.go

```diff
@@ -0,0 +1,164 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/ipfs/testground/pkg/tgwriter"
+	"github.com/mitchellh/mapstructure"
+)
+
+// Client is the API client that performs all operations
+// against a testground server.
+type Client struct {
+	// client used to send and receive http requests.
+	client   *http.Client
+	endpoint string
+}
+
+// New initializes a new API client
+func New(endpoint string) *Client {
+	return &Client{
+		client:   &http.Client{},
+		endpoint: endpoint,
+	}
+}
+
+// Close the transport used by the client
+func (c *Client) Close() error {
+	if t, ok := c.client.Transport.(*http.Transport); ok {
+		t.CloseIdleConnections()
+	}
+	return nil
+}
+
+// List sends `list` request to the daemon.
+// The Body in the response implement an io.ReadCloser and it's up to the caller to
+// close it.
+func (c *Client) List(ctx context.Context) (io.ReadCloser, error) {
+	return c.request(ctx, "GET", "/list", nil)
+}
+
+// Describe sends `describe` request to the daemon.
+// The Body in the response implement an io.ReadCloser and it's up to the caller to
+// close it.
+func (c *Client) Describe(ctx context.Context, r *DescribeRequest) (io.ReadCloser, error) {
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.request(ctx, "GET", "/describe", bytes.NewReader(body.Bytes()))
+}
+
+// Build sends `build` request to the daemon.
+// The Body in the response implement an io.ReadCloser and it's up to the caller to
+// close it.
+func (c *Client) Build(ctx context.Context, r *BuildRequest) (io.ReadCloser, error) {
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.request(ctx, "POST", "/build", bytes.NewReader(body.Bytes()))
+}
+
+// Run sends `run` request to the daemon.
+// The Body in the response implement an io.ReadCloser and it's up to the caller to
+// close it.
+func (c *Client) Run(ctx context.Context, r *RunRequest) (io.ReadCloser, error) {
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.request(ctx, "POST", "/run", bytes.NewReader(body.Bytes()))
+}
```

Comment:
So, we always send JSON to the server and receive a streaming response of only bytes? Are these bytes base64 encoded and encoded in JSON as well?

I'm asking this as I'm considering future browser clients / dashboards etc. We've had to tackle many problems in JS land because Go often assumes that everything is a TCP socket and makes things super painful to deal with in a Browser or JS contexts (some notes from those wars can be found at https://github.com/ipfs/interface-js-ipfs-core/issues/81)

Reply:
We have two types of API endpoints - those that immediately return a response (for example `testground list`), and those that stream response back to the client (for example `testground build` and `testground run`).

Currently all requests/responses are JSON marshalled.

For the streaming endpoints, we use chunked HTTP responses and return JSON encoded `Msg`s (see PR description). I will review the linked issue, as I am not sure what could be a potential problem with this protocol.

I basically copied whatever Docker client/server was doing, because the model seemed like a good fit for `testground`.